### DISCUSSION TITLE
select and record different audio versions

### DISCUSSION THREAD

**Discussion by dverdier** [aclap-dev/video-downloadhelper#956] [Help and Q&A] [answered]
### Your question

Hello and thank you for your help. I'm watching a video with several language options.

example : [https://www.arte.tv/fr/videos/004899-000-A/rome-ville-ouverte/](url)

My problem: VDH automatically saves only one version but I cannot save the other versions (even if I change language while viewing). 

I work with Mac PS Ventura Version 13.6.7 (22G720)

and I use the latest version of firefox mozilla. 

### Addon & CoApp version and details

version: 9.0.2.12
target: mozilla
channel: stable
lang: fr
coapp: {"found":true,"path":"/Applications/net.downloadhelper.coapp.app/Contents/MacOS/vdhcoapp","version":"2.0.19","new_version":false}
license: {"accepted":true,"email":"<removed>","key":"ca39187217cd4f26"}
platform: x86-64 mac
UA: Mozilla/5.0 (Macintosh; Intel Mac OS X 10.15; rv:127.0) Gecko/20100101 Firefox/127.0
{
    "valid_license_message_has_been_displayed": true,
    "used_sidebar_button": true,
    "download_directory": "/Users/verdier/dwhelper",
    "last_download_directory": {
        "__serializer_tag": "option_some",
        "__serializer_value": {
            "__serializer_tag": "primitive"
        }
    },
    "view_options": {
        "all_tabs": false,
        "low_quality": false,
        "sort_by_status": true,
        "sort_reverse": false,
        "show_button_clean": true,
        "show_button_clean_all": false,
        "show_button_convert_local": false,
        "hide_downloaded": false
    },
    "use_sidebar": true,
    "open_count_store": 3,
    "successfull_dl": 9
}

**Comment by GrampaWildWilly** [comment]
I do not work for VDH.  I am just another user.

![#01](https://github.com/aclap-dev/video-downloadhelper/assets/75588964/894d61ab-5b7e-42e0-abec-947ab0231363)

Domage.  J'aurais bien aimer chercher une solution pour toi.  J'imagine que c'est possible de trouver toutes les bandes sonores disponibles mais je crois qu'il faut piger dans le Network Monitor, cette chose:

![image](https://github.com/aclap-dev/video-downloadhelper/assets/75588964/47502808-95ec-4557-9cff-bd93dd42a14f)

J'aurais souhaiter trouver une façon de télécharger des bandes alternatives avec ffmpeg.  Tu peux trouver des exemples de comment faire tout ça en faisant la recherche `Plan B` dans ce forum sur GitHub.  Je ne suis pas convaincu à 100% que cela réussirait à te donner les autres bandes sonores, mais on ne sait jamais sans essayer.   Ça vaut la peine d'essayer.

**Reply by dverdier** [reply]
Thank you for your answer. Please, could you try try with another movie link available worldwide (with different soundtracks) ?

**Reply by GrampaWildWilly** [reply]
You will have to give me such a link.  Make it easy for me to help you.  Give me a URL for a video & I will try it.

**Reply by GrampaWildWilly** [reply]
OK.  I just went to the arte.tv home page & picked the first thing on the page.  It happened to lead me to this page:

https://www.arte.tv/en/videos/118869-000-A/arte-reportage/

Luckily, this is a video with the right properties.  First, I had to launch playback.

![#01](https://github.com/aclap-dev/video-downloadhelper/assets/75588964/ccea0be2-9594-4533-864f-82db1a8aa150)
![#02](https://github.com/aclap-dev/video-downloadhelper/assets/75588964/63a9558a-8a8a-4ca8-8dd7-5eb0e527a7b9)

So I opened the Network Monitor & reloaded the page.  That got me exactly what I was hoping to find: an HLS master manifest.  That's what is highlighted in this next image.  The type of the manifest is x-mpegurl.  That is partially visible in the image under the Type column.

![#03](https://github.com/aclap-dev/video-downloadhelper/assets/75588964/83126801-13f1-46fb-8647-deeb2ec89ad2)

I popped up the context menu on that entry & extracted its URL using the menu entry you can see in that image.  I pasted that into my ffprobe .bat file & executed this command:

"C:\DownloadHelper CoApp\ffprobe.exe"  -protocol_whitelist file,crypto,data,http,https,tls,tcp -strict experimental -analyzeduration 25000000 -probesize 250000000 -allowed_extensions ALL "https://manifest-arte.akamaized.net/api/manifest/v1/Generate/202406051080B9DFAE5B8D7FB77BAB13F645F08B20/en/XQ+KS/118869-000-A.m3u8" 2>"Q:\To Watch\ffprobe.txt"

I can't tell you how to do this on Mac.  I have utterly no experience with Mac.  I am doing this on Windows & this is completely routine command execution at a command prompt.  I am executing the ffprobe that is packaged with the CoApp.  To simplify the structure of the command, think of it like this:

ffprobe [parameters] URL 2>CaptureFile

The URL is the one I extracted from the Network Monitor.  It is the URL of the HLS master manifest.  This URL will be different for each visitor to that web page.  So you CANNOT use the URL I'm showing here.  You have to visit the page & follow the same steps I did.  Your URL will probably look a lot like mine but it will not be identical.

The 2> thing on the end is a Windows tool for capturing the output of the command in a file.  It's called redirection in Microsoft jargon.  If you don't supply that, the output just scrolls by in the command window & you won't be able to inspect it at your leisure.  I'm relying on you to know how to do the analogous thing on Mac.  The file I captured is this:

[ffprobe.txt](https://github.com/user-attachments/files/16075129/ffprobe.txt)

It may take you a bit to get used to the way data is presented in these listings.  Look for the lines that say Program 0, Program 1, & so on.  Those are the individual videos available here.  You can see that each one consists of a number of items labelled Stream.  Each Stream has an identifier of the form 0:1, 0:2, etc.  The one we're interested in is Program 1 Stream 0:1.  That is the 1920x1080 video track.  It conveniently contains the word Video in its information.

Below the video Stream, you can see a number of Streams that are audio tracks, each of a different language.  Once again, it conveniently contains the word Audio in its information.  The language of the audio track is indicated by the Stream identifier being something like x:y(ll), where ll is fr for French, de for German, en for English, & so on.  What you see here matches up with what I'm showing in my second image above.

Below the audio streams, you can see the caption Streams.  Subtitle Streams.  Whichever you want to call them.  Caption & subtitle are synonyms.  They are all in WebVTT format.  Once again, the language of each caption stream is indicated in the Stream identifier.  What you see here also matches up with what I'm showing in my second image above.

For our video track, we want Stream 0:1.  For our audio track, let's pick Stream 0:8, the French audio track.  For our subtitles, let's pick Stream 0:10, the English captions.  Now supposedly, ffmpeg can embed captions within a video.  I've tried it & it doesn't work.  I probably didn't do it right but there are advantages to keeping the captions in a separate file.  For one thing, you can edit them if you discover something like a typo or even a bad translation.  It's no hardship to keep the subtitles in a separate file.  So, using the information provided by the ffprobe report, I constructed a command to execute ffmpeg.  Twice.  Once for the video with its audio, once for the subtitles.  Here are the results.  Each log file shows the exact command I executed near the top.  At the completion of the ffmpeg command, I have captured a few statistics, like the size of the output file, the duration of the download, & the speed of the download.  That is followed by ffprobe on the output file.

[ARTE Test Video mp4 Log.txt](https://github.com/user-attachments/files/16075132/ARTE.Test.Video.mp4.Log.txt)
[ARTE Test Video vtt Log.txt](https://github.com/user-attachments/files/16075134/ARTE.Test.Video.vtt.Log.txt)

The video downloaded fine, as you can see.  But there's some kind of problem with the subtitles.  The log file you see there is for a second attempt to get subtitles from Stream 0:18, which is supposedly the multilingual default.  But it's getting the same error I got with the 0:10 English captions.  I don't know what's wrong there.  And I have to say I am not sufficiently motivated to investigate this any further.  My initial reaction is that there is some sort of error on the web site that is blocking me from downloading the subtitles.  Maybe another video wouldn't have the same problem.  Maybe you have to actually select one of the subtitle languages & get them to display in the player before meaningful information appears in the Netwrork Monitor.  I'll leave that for you to figure out.

In any case, this shows how you would get whichever audio track you want.  Just pick a different Stream identifier for the audio track in your preferred language.

Wait.  I looked at the Network Monitor more closely.  There is an entry in there for something of type vtt.  It is a caption file, apparently in English.

![#04](https://github.com/aclap-dev/video-downloadhelper/assets/75588964/3f0cba03-4b4e-496f-bd4c-38f884a2b8f8)

So I launched playback again & opened the tools menu again.  I selected another language for the subtitles.

![#05](https://github.com/aclap-dev/video-downloadhelper/assets/75588964/d1a4dd1e-dd24-4a85-adb4-9aad6bfc5a50)

You can just double click any entry in the Network Monitor & Firefox will open that object in its own browser tab.

![#06](https://github.com/aclap-dev/video-downloadhelper/assets/75588964/db6202ee-f22b-4d3e-b161-d3893fec50b3)

Just highlight the full contents of that browser tab & copy it to your system clipboard.  Then paste that into a text editing session.  Then save the file using the same name as the MP4 you just downloaded, but give it a file extension of .vtt instead of .mp4.  Here I'm playing it in VLC:

![#07](https://github.com/aclap-dev/video-downloadhelper/assets/75588964/37ec8adb-4af6-4936-97a3-14ddd84b0971)

You can see it's playing fine & I've got Spanish subtitles.  You'll have to take my word for it that the audio was also fine & it was indeed French.

One interesting thing I notice with these captions is that they start with several formatting directives controlling colors.  Each subtitle contains directives for size & position.  I don't think VLC handles these correctly.  I have a caption editing application (totally freeware) called SubtitleEdit, affectionately known as SE.  I played this video in SE & I saw some subtitles that VLC did not display.  There are some choices for video player in SE.  It can use VLC.  On advice from Nik (the author of SE), I am using MPV as the video player in SE.  There are Spanish subtitles translating the text message session near the start of this video.  None of those appeared in VLC, but SE showed them.  I didn't inspect the full video closely enough to tell whether any color, size, or position directives were being better obeyed in SE than in VLC.  Just a heads up that VLC is not always the best video player.

If you want to do all this yourself, you're going to have to spend some time learning how to execute these commands in the Mac environment.  Like I said, I can't help you with that since I am totally ignorant of the Mac environment.  Plus, you're going to have to learn a bit about ffmpeg.  I can't teach you everything.  You'll have to put in some effort yourself.  I have no idea if supporting this sort of video is even in the plans for VDH.

**Reply by GrampaWildWilly** [reply]
You might find this helpful:

https://groups.google.com/g/video-downloadhelper-q-and-a/c/oYDFx367MSY

**Reply by dverdier** [reply]
Thank you very much for your very long and detailed reply. Unfortunately, I'm unable to find a video file in m3u8 format on my Mac (even if I follow your advice to the letter). For example, here are several mp4 files that appear on this video :

https://www.arte.tv/fr/videos/022054-000-A/mariage-a-l-italienne/

they mention an ‘Italian’ version but when I use VDH, the video is automatically recorded in French. With older versions of VDH, the different audio versions would appear when you regenerated the page. Alternatively, it was possible to record silent video and audio separately... with the option of merging the two afterwards. Why is the new version of VDH less functional than its predecessor? 

<img width="1399" alt="Capture d’écran 2024-08-27 à 17 00 46" src="https://github.com/user-attachments/assets/444a73a6-a07a-4fd9-a39b-79ba4fac4972">

**Reply by GrampaWildWilly** [reply]
I can't investigate that video.  It is geoblocked for me.  As for what used to work & no longer works in VDH, I can't answer that.  I don't work for VDH.  I'm just another user.

**Reply by GrampaWildWilly** [reply]
By the way, I did not get anything useful in the Network Monitor until I launched playback & paused it after a couple of seconds.  I say that right at the beginning of my post above from July 2.  You need to let the video play for a couple of seconds in the web page before the useful information appears in the Network Monitor.  The portion of the Network Monitor that you are showing in your image is near the middle of the information.  The manifest would appear near the bottom.  Be sure you scroll the Network Monitor & look carefully at everything that's in there before you conclude something isn't there.

**Comment by Mattmjs** [comment]
I'm geoblocked on that film but used france vpn extension to see it. I think the problem is you've got too many things selected in the
network. And that makes it harder to find what you're after. You've got HTML, CSS, JS, XHR & Media in your image.
Have it selected to All and filter the network for m3u8 or select it to either XHR or Media. I think filtering the network is the best option.

You can get the video and audio separately by opening the video and audio stream manifests in a new tab by looking at my images.


The m3u8 on top is the master manifest, it has resolutions from 1080p to 216p but it seems the video player has no switch to
choose your video quality. Anyway the one below I highlight is the playlist for 1080p. Other versions might show 720.m3u8 or 360.m3u8
Open this in a new tab, there might be a prompt that comes up but dismiss or cancel it. Turn your attention to the VDH in the toolbar.
It'll detect it, if you have it set to Download it will give you a default name like this: arte-cmafhls akamaized netamcmaf022000022000022054-000-C20240824. If you have it set to Download as you can give it proper name before download starts.
![italian film](https://github.com/user-attachments/assets/59d836d5-ecd4-4425-a7df-2a86cc0ae252)


Here is the italian audio I've highlighted, just follow the same steps to save it like the video. You then should have both video & audio
which you'll need to merge together or play the files together in VLC.

![italian audio](https://github.com/user-attachments/assets/56983747-92ed-4592-befe-0b96bbed9155)

**Reply by dverdier** [reply]
Once again: many thanks for your answers. I followed the procedure and let the video run for a good ten minutes. I sorted the formats and this is what I get: 
<img width="1370" alt="Capture d’écran 2024-08-28 à 14 18 55" src="https://github.com/user-attachments/assets/cba916d9-b642-4c9c-ac27-51c803767f23">

**Reply by dverdier** [reply]
What should I do with these files? I've saved them on my hard drive but merging them with VLC seems very complicated...

**Reply by Mattmjs** [reply, marked as answer]
I don't think you can merge files together in VLC. What I mean is VLC will play the separate files together synchronously.
But if you would rather a single file, you'd need some software to do that. VDH has this option in the legacy interface.

**Reply by dverdier** [reply]
You are totally right. I know that VDH had this merging option but is this still the case ?

**Reply by GrampaWildWilly** [reply]
In discussion #466 in a post on May 11, Paul says he has purposely NOT ported the old `Aggregate local audio and video files` logic forward to VDH 9.  The option is in the legacy UI but it doesn't do anything.  However, merging is so easy to do with ffmpeg.  I commented on that in #466.

**Reply by akostadinov** [reply]
FYI I have a script [1] that would mux files in a directory with subtitles. I haven't considered multi-language audio though. I guess it can be added similar to how subtitles are added. I don't see an option in VDH to select a specific audio stream. Neither in new or old interface. Seems to try to make it an mp3 which is unthinkable.

https://gist.github.com/akostadinov/8197947def1fe61c12a5ffffc615c969

**Reply by Mattmjs** [reply]
Even I can be wrong on things sometimes, you can actually merge files together in VLC. I saw a video on youtube and applied
those steps and it did what was shown.
